src/context/typecore.ml

```diff
@@ -222,7 +222,31 @@ let add_local ctx k n t p =
 	ctx.locals <- PMap.add n v ctx.locals;
 	v
 
+let check_identifier_name ctx name kind p =
+	if starts_with name '$' then
+		display_error ctx (kind ^ " names starting with a dollar are not allowed") p
+	else
+		let is_valid_identifier s =
+			match s with
+			| "new" | "this" -> true
+			| _ -> Lexer.is_valid_identifier s
+		in
+		if not (is_valid_identifier name) then display_error ctx ("`" ^ name ^ "` is not a valid " ^ kind ^ " name") p
+
+let check_local_variable_name ctx name origin p =
+	let s_var_origin origin =
```

Comment:
Why function if called immediately and one time only?

Reply:
Originally I wanted to have it in a more general place because it seems it could be useful elsewhere, but wasn't sure where to best put it.

Reply:
Note that this pattern is fine in OCaml, it should be inlined properly and is cleaner. Also allows to move out the code elsewhere at some point.

Reply:
Somewhere in `type.ml` where `tvar_origin` is declared.